spec/unit/prepare.spec.js

```diff
@@ -0,0 +1,935 @@
+
+        cordovaProject = {
+            root: '/mock',
+            projectConfig: {
+                path: '/mock/config.xml',
+                cdvNamespacePrefix: 'cdv'
+            },
+            locations: {
+                plugins: '/mock/plugins',
+                www: '/mock/www'
+            }
+        };
+        platformResourcesDir = PATH_RESOURCE;
+
+        emitSpy = jasmine.createSpy('emit');
+        prepare.__set__('events', {
+            emit: emitSpy
+        });
+
+        updatePathsSpy = jasmine.createSpy('updatePaths');
+        prepare.__set__('FileUpdater', {
+            updatePaths: updatePathsSpy
+        });
+
+        // mocking initial responses for mapImageResources
+        prepare.__set__('mapImageResources', function (rootDir, subDir, type, resourceName) {
+            if (resourceName.includes('ic_launcher.png')) {
+                return createResourceMap('ic_launcher.png');
+            } else if (resourceName.includes('ic_launcher_foreground.png')) {
+                return createResourceMap('ic_launcher_foreground.png');
+            } else if (resourceName.includes('ic_launcher_background.png')) {
+                return createResourceMap('ic_launcher_background.png');
+            } else if (resourceName.includes('ic_launcher_foreground.xml')) {
+                return createResourceMap('ic_launcher_foreground.xml');
+            } else if (resourceName.includes('ic_launcher_background.xml')) {
+                return createResourceMap('ic_launcher_background.xml');
+            } else if (resourceName.includes('ic_launcher.xml')) {
+                return createResourceMap('ic_launcher.xml');
+            }
+        });
+    });
+
+    it('Test#001 : Should detect no defined icons.', function (done) {
+        const updateIcons = prepare.__get__('updateIcons');
+
+        // mock data.
+        cordovaProject.projectConfig.getIcons = function () {
+            return [];
+        };
+
+        updateIcons(cordovaProject, platformResourcesDir);
+
+        // The emit was called
+        expect(emitSpy).toHaveBeenCalled();
+
+        // The emit message was.
+        let actual = emitSpy.calls.argsFor(0)[1];
+        let expected = 'This app does not have launcher icons defined';
+        expect(actual).toEqual(expected);
+
+        done();
+    });
+
+    it('Test#002 : Should detech incorrect configrations for adaptive icon and throws error.', function (done) {
+        const updateIcons = prepare.__get__('updateIcons');
+
+        // mock data.
+        cordovaProject.projectConfig.getIcons = function () {
+            return [{
+                src: undefined,
+                target: undefined,
+                density: 'mdpi',
+                platform: 'android',
+                width: undefined,
+                height: undefined,
+                background: 'res/icon/android/mdpi-background.png',
+                foreground: undefined
+            }];
```

Comment:
You have something similar to this in almost every test. Might be nice to factor this out e.g as a function that fills in missing props as undefined. This could help to slim down this 1000 lines spec.